codec/src/main/java/io/netty/handler/codec/MessageToMessageEncoder.java

```diff
@@ -108,28 +109,36 @@ public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise)
             if (out != null) {
                 final int sizeMinusOne = out.size() - 1;
                 if (sizeMinusOne == 0) {
-                    ctx.write(out.get(0), promise);
+                    ctx.write(out.getUnsafe(0), promise);
                 } else if (sizeMinusOne > 0) {
                     // Check if we can use a voidPromise for our extra writes to reduce GC-Pressure
                     // See https://github.com/netty/netty/issues/2525
-                    ChannelPromise voidPromise = ctx.voidPromise();
-                    boolean isVoidPromise = promise == voidPromise;
-                    for (int i = 0; i < sizeMinusOne; i ++) {
-                        ChannelPromise p;
-                        if (isVoidPromise) {
-                            p = voidPromise;
-                        } else {
-                            p = ctx.newPromise();
-                        }
-                        ctx.write(out.getUnsafe(i), p);
+                    if (promise == ctx.voidPromise()) {
+                        writeVoidPromise(ctx, out);
+                    } else {
+                        writePromiseCombiner(ctx, out, promise);
                     }
-                    ctx.write(out.getUnsafe(sizeMinusOne), promise);
                 }
                 out.recycle();
             }
         }
     }
 
+    private static void writeVoidPromise(ChannelHandlerContext ctx, CodecOutputList out) {
+        final ChannelPromise voidPromise = ctx.voidPromise();
+        for (int i = 0; i < out.size(); i++) {
```

Comment:
why are we using out.size() here when the original code used sizeMinusOne?

Reply:
oh right it loops the first bunch then writes the last..

Reply:
good catch! @bryce-anderson please fix and adjust test to catch this